### PR TITLE
Extract Assignments from ProjectNode and ApplyNode

### PR DESCRIPTION
**Pull request by kokosing**
Extract Assignments from ProjectNode and ApplyNode

Assignments is a map of Symbol to Expression. It represents a domain
object which is used quite commonly in the planner. This commit extract
it as separate entity which bunch of utility methods.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
+        this.outputs = ImmutableList.copyOf(assignments.keySet());
+    }
+
+    public List<Symbol> getOutputs()
+    {
+        return outputs;
+    }
+
+    public Map<Symbol, Expression> getAssignments()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: `assignments.getAssignments()`?

Maybe rename to `asMap()` or `toMap()`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    public Assignments rewrite(Function<Expression, Expression> rewrite)
+    {
+        return assignments.entrySet().stream()
+                .map(entry -> Maps.immutableEntry(entry.getKey(), rewrite.apply(entry.getValue())))
+                .collect(toAssignments());
+    }
+
+    public Assignments filter(Collection<Symbol> symbols)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: make also `filter` method that accepts `predicate`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java [hunk 3]
```diff
@@ -128,19 +128,16 @@ private Optimizer(PlanNodeIdAllocator idAllocator)
                     node.getHashSymbol(),
                     Optional.empty());
 
-            ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
-            for (Symbol symbol : aggregationNode.getOutputSymbols()) {
-                outputSymbols.put(symbol, symbol.toSymbolReference());
-            }
+            Assignments assignments = Assignments.identity(aggregationNode.getOutputSymbols());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SingleDistinctOptimizer.java]: use builder instead? Then you don't need `append` method

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                Assignments::builder,
+                (builder, entry) -> builder.put(entry.getKey(), entry.getValue()),
+                (left, right) -> {
+                    left.putAll(right.build());
+                    return left;
+                },
+                Assignments.Builder::build);
+    }
+
+    public Assignments append(Symbol symbol, Expression expression)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: maybe remove this method? There is just one usage with can be rewritten to using builder. I think builder is convention in presto.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        return builder().putAll(this).put(symbol, expression).build();
+    }
+
+    public Collection<Expression> getExpressions()
+    {
+        return assignments.values();
+    }
+
+    public Set<Symbol> getSymbols()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: outputs and symbols should be the same I think

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        return builder().put(symbol, expression).build();
+    }
+
+    public static Assignments of(Symbol symbol1, Expression expression1, Symbol symbol2, Expression expression2)
+    {
+        return builder().put(symbol1, expression1).put(symbol2, expression2).build();
+    }
+
+    private final Map<Symbol, Expression> assignments;
+    private final List<Symbol> outputs;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: Class is immutable so there is no need to make them in sync. They are always in sync.

map returns a set for keys, so it has to be copied

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: I renamed that to `getMap` as there is no conversion (it just a getter) so prefix `to` or `as` could be misleading

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Assignments.java]: they have different return type

**Comment by kokosing** [comment]
This pull request was merged as part of another pull request, see https://github.com/prestodb/presto/pull/6808

**Pull request closed without merging**